### ISSUE TITLE
UI API / Feature Gaps

### ISSUE THREAD

**Issue by gaasedelen** [Vector35/binaryninja-api#1097]
I promised @psifertex to make an issue with some of the API/features I felt I was missing while developing the frontend [Ethrays](http://blog.ret2.io/2018/05/16/practical-eth-decompilation/) integration. These will probably be mostly relevant to developing UI plugins.

- [x] 1. It would be nice to have some API's to open (or close) a BNDB/Binary into the binja UI, creating a new tab/session (or whatever you want to call it)

```
open_bndb(filepath)         # open existing bndb from disk as a new binja tab/session
close_bndb(...)             # close a bndb/file as if a user had 'X'd out its tab in binja
open_binary(filepath, ...)  # open 'foo.dll' from disk as if loaded through UI
...
```

- [x] 2. A non-hacky way to access the current UI state of 'what the user is looking at'

```
get_current_address()
get_current_func()
get_current_bv()
...
```

- [ ] 3. I get that background tasks are a first-class citizen of BN, but sometimes it really might just be better to pop a blocking waitbox. 

```
show_wait_box(...)
hide_wait_box(...)
```

- [x] 4. I should be able to create a custom BN 'view'. Also, the ability to switch available views. For example, I wanted to create a 'Decompilation View' for my plugin (versus the existing assembly, hex, or IL views). It is not possible for a user to do this. 

EDIT: I'm going to consider this done at this point.. at least the binja UI has the 'widget creation' features necessary for plugins like lighthouse.

```
blank_view_widget = create_view(...) # I want a blank qwidget I can put stuff in...
switch_view(...) 
... 
```

- [x] 5. There should be an API to get visibility/access into all open 'sessions' (bvs open in the ui).

```
sessions = get_all_sessions()
```

- [x] 6. There are some events / callbacks I would have liked to have had for Ethrays:

```
bv_created(bv)  # bv / binja session
bv_closing(bv)

current_address_changed(bv, address)
current_function_changed(bv, func)
current_view_changed(bv)

binja_closing()

... probbably a few others
``` 

I started noting these down only towards the end of development, so this list is probably missing a few other gripes I had.

**Comment by plafosse** [comment]
A lot of these we believe will be addressed by #305

**Comment by gaasedelen** [comment]
I believe most if not all of these have been added to the binja API this point, it's probably safe to close \o/

If there are any similar requests/missing API's others can open them as new issues.
